internal/ethapi/tracer.go

```diff
@@ -240,21 +232,20 @@ func NewJavascriptTracer(code string, timeout time.Duration) (*JavascriptTracer,
 		db:         db,
 		dbvalue:    db.toValue(vm),
 		err:        nil,
-		timeout:    timeout,
 	}, nil
 }
 
+// Stop terminates execution of any JavaScript
+func (jst *JavascriptTracer) Stop(err error) {
+	jst.err = err
+	jst.vm.Interrupt <- func() {
+		panic(err)
+	}
+}
```

Comment:
The assignment of jst.err is racy because it is also assigned in callSafely.
IMHO it would be simpler to just set the vm cancel channel to ctx.Done.

To get the right error, you can check for jst.err after tracing and return it when non-nil.
If jst.err is non-nil, just return ctx.Err which will be non nil if the context was canceled.